doc(pe): Document policy options (#601)
armory/docs: content/en/docs/plugin-guide/plugin-policy-engine/packages/spinnaker.execution/stages.before/deployManifest.md

---
title: "spinnaker.execution.stages.before.deployManifest"
linktitle: "deployManifest"

> description: "A policy targeting this object is run before executing each task in a deployManifest stage."

Reviewer: ```suggestion
description: "A policy targeting this object runs before executing each task in a Deploy Manifest stage."
```

          "startTime": null,
          "status": "NOT_STARTED"
        }
      ],
      "type": "deployManifest"
    },
    "user": {
      "isAdmin": false,
      "roles": [],
      "username": "myUserName"
    }
  }
}
```
</details>

## Example Policy

- This policy requires that a set of annotations have been applied to any manifests that are being deployed. Specifically the annotations 'app' and 'owner' must have been applied.

  {{< prism lang="rego" line-numbers="true" >}}
    package spinnaker.execution.stages.before.deployManifest
    required_annotations:=["app","owner"]
    deny["Manifest is missing a required annotation"] {
        metadata :=input.stage.context.manifests[_].metadata
        annotations := object.get(metadata,"annotations",{})
        # Use object.get to check if data exists
        object.get(annotations,required_annotations[_],null)==null
    }{
        metadata :=input.stage.context.manifests[_].spec.template.metadata
        annotations := object.get(metadata,"annotations",{})
        # Use object.get to check if data exists
        object.get(annotations,required_annotations[_],null)==null
    }
  {{< /prism >}}

<br/>

> - This policy requires prevents exposing a set of ports that are unencrypted buy have encrypted alternatives. Specifically this policy prevents exposing HTTP, FTP, TELNET, POP3, NNTP, IMAP, LDAP, and SMTP from a pod, deployment, or replicaset.

Reviewer: ```suggestion
- This policy requires prevents exposing a set of ports that are unencrypted buy have encrypted alternatives. Specifically, this policy prevents exposing HTTP, FTP, TELNET, POP3, NNTP, IMAP, LDAP, and SMTP from a pod, deployment, or replica set.
```

Reviewer: replica set for the generic name of the thing. K8s specifically calls it ReplicaSet if you want to use that

  {{< prism lang="rego" line-numbers="true" >}}
  package spinnaker.execution.stages.before.deployManifest

  blockedPorts := [20,21,23,80,110,119,143,389,587,8080,8088,8888]

  deny["A port typically used by an unencrypted protocol was detected."] {
      #Check for service
      ports := input.stage.context.manifests[_].spec.ports[_]
      any([object.get(ports,"port",null) == blockedPorts[_], 
            object.get(ports,"targetPort",null) == blockedPorts[_]])
  }{ 
      #Check for pod
      input.stage.context.manifests[_].spec.containers[_].ports[_].containerPort=blockedPorts[_]
  } { 
      #Check for pod template
      input.stage.context.manifests[_].spec.template.spec.containers[_].ports[_].containerPort=blockedPorts[_]
      }
  {{< /prism >}}

<br/>

> - This policy checks whether or not the image being approved is on a list of imaged that are approved for deployment. The list of what images are approved must seperately be uploaded to the OPA data document

Reviewer: ```suggestion
- This policy checks whether or not the image being approved is on a list of imaged that are approved for deployment. The list of what images are approved must seperately be uploaded to the OPA data document.
```

  {{< prism lang="rego" line-numbers="true" >}}
  package spinnaker.execution.stages.before.deployManifest

  deny["Manifest creates a pod from an image that is not approved by the security scanning process."] {
  #Check pod templates
      isImageUnApproved(input.stage.context.manifests[_].spec.template.spec.containers[_].image)
  } {#check pods
      isImageUnApproved(input.stage.context.manifests[_].spec.containers[_].image)
  }
  {#check pod template initContainers
      isImageUnApproved(input.stage.context.manifests[_].spec.template.spec.initContainers[_].image)
  }
  {#check pod initContainers
      isImageUnApproved(input.stage.context.manifests[_].spec.initContainers[_].image)
  }

  isImageUnApproved(image){    not isImageApproved(image) }
  isImageApproved(image){    image==data.approvedImages[_]}
  {{< /prism >}}

<br/>

> - This policy prevents applications from deploying to namespaces that they are not whitelisted for.

Reviewer: ```suggestion
- This policy prevents applications from deploying to namespaces that they are not allowed to.
```

Reviewer: inclusive language

      allowedNamespaces[i].ns[_]==namespace
  }
  {{< /prism >}}

## Keys

Parameters related to the stage against which the policy is executing can be found in the [input.stage.context](#inputstagecontext) object.

### input.pipeline

| Key                                               | Type      | Description                                                                                                                                                          |
| ------------------------------------------------- | --------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| `input.pipeline.application`                      | `string`  | The name of the application to which this pipeline belongs.                                                                                                          |
| `input.pipeline.authentication.allowedAccounts[]` | `string`  | The list of accounts to which the user this stage is running as has access.                                                                                          |

> | `input.pipeline.authentication.user`              | `string`  | The Spinnaker user initiating the change.                                                                                                                            |

Reviewer: ```suggestion
| `input.pipeline.authentication.user`              | `string`  | The user initiating the change.                                                                                                                            |
```

| `input.pipeline.partition`                        |           |                                                                                                                                                                      |
| `input.pipeline.paused`                           |           |                                                                                                                                                                      |
| `input.pipeline.pipelineConfigId`                 |           |                                                                                                                                                                      |
| `input.pipeline.source`                           |           |                                                                                                                                                                      |
| `input.pipeline.spelEvaluator`                    | `string`  | Which version of spring expression language is being used to evaluate SpEL.                                                                                          |

> | `input.pipeline.stages[]`                         | `string`  | An array of the stages in the pipeline. Typically if you are writing a policy that examines multiple pipeline stages, it is better to write that policy against either the `opa.pipelines` package, or the `spinnaker.execution.pipelines.before package`. |

Reviewer: ```suggestion
| `input.pipeline.stages[]`                         | `string`  | An array of the stages in the pipeline. Typically if you are writing a policy that examines multiple pipeline stages, it is better to write that policy against either the `opa.pipelines` package or the `spinnaker.execution.pipelines.before package`. |
```

| `input.stage.context.kato.task.firstNotFoundRetry`   | `number`   |                                                                                                                                                                                      |
| `input.stage.context.kato.task.lastStatus`           | `string`   |                                                                                                                                                                                      |
| `input.stage.context.kato.task.notFoundRetryCount`   | `number`   |                                                                                                                                                                                      |
| `input.stage.context.kato.task.terminalRetryCount`   | `number`   |                                                                                                                                                                                      |

> | `input.stage.context.kato.tasks[]`                   | `string`   | This array contains the tasks that are executed during this stage. when each task finishes execution, the policy will be rechecked with the status of each execution task updated.   |

Reviewer: ```suggestion
| `input.stage.context.kato.tasks[]`                   | `string`   | This array contains the tasks that are executed during this stage. When each task finishes execution, the policy will be rechecked with the status of each execution task updated.   |
```

| `input.stage.context.kato.tasks[].resultObjects[]`   | `{object}` | DO NOT USE. This contains numerous manifests/artifacts created by the execution of the stage. Typically policies should be written against the inputs to the stage, not its outputs. |
| `input.stage.context.kato.tasks[].status.completed`  | `boolean`  | True when the task has finished executing.                                                                                                                                           |
| `input.stage.context.kato.tasks[].status.failed`     | `boolean`  | Did the task attempt to execute and fail.                                                                                                                                            |
| `input.stage.context.kato.tasks[].status.retryable`  | `boolean`  | Can the task retry if execution fails.                                                                                                                                               |

> | `input.stage.context.manifestArtifactAccount`        | `string`   | What artifact account are artifacts stored in.                                                                                                                                       |

Reviewer: ```suggestion
| `input.stage.context.manifestArtifactAccount`        | `string`   | The artifact account are artifacts stored in.                                                                                                                                       |
```

> | `input.stage.context.manifests[]`                    | `object`   | The raw kubernetes manifest being deployed. This is the element you should write policies against that involve the manifest.                                                         |

Reviewer: ```suggestion
| `input.stage.context.manifests[]`                    | `object`   | The raw Kubernetes manifest being deployed. This is the element you should write policies against that involve the manifest.                                                         |
```

| `input.stage.context.moniker.cluster`                | `string`   | The name of the cluster to which this stage is deploying.                                                                                                                            |
| `input.stage.context.outputs`                        | `object`   | DO NOT USE. This contains numerous manifests/artifacts created by the execution of the stage. Typically policies should be written against the inputs to the stage, not its outputs. |
| `input.stage.context.source`                         | `string`   | Specifies whether the manifest is coming from an artifact or was directly specified in the pipeline as text.                                                                         |
| `input.stage.context.stableManifests[].location`     | `string`   | The namespace of the stable manifest.                                                                                                                                                |

> | `input.stage.context.stableManifests[].manifestName` | `string`   | The manifests name.                                                                                                                                                                  |

Reviewer: ```suggestion
| `input.stage.context.stableManifests[].manifestName` | `string`   | The manifest's name.                                                                                                                                                                  |
```
